tutorials/data-science-preprocessing/index.md

```diff
@@ -0,0 +1,437 @@
+* For running the pipeline in the cloud, you also must enable the [Cloud
+  Dataflow, Compute Engine, Cloud Logging, Cloud Storage, and Cloud Storage
+  JSON APIs][dataflow-apis].
+
+[setup]: /getting-started#set_up_a_project
+[python]: https://www.python.org/
+[pip]: https://pip.pypa.io/en/latest/installing/
+[virtualenv]: http://virtualenv.pypa.io
+[bq-api]: https://console.cloud.google.com/flows/enableapi?apiid=bigquery
+[dataflow-apis]: https://console.cloud.google.com/flows/enableapi?apiid=dataflow,compute_component,logging,storage_component,storage_api,bigquery
+
+## Download the tutorial data
+
+The [Meteorite Landings
+API](https://catalog.data.gov/dataset/meteorite-landings-api) provides
+information in JSON format from The Meteoritical Society on all known meteorite
+landings. You can download the first page of 1000 objects (according to the
+[developer documentation](https://dev.socrata.com/docs/paging.html)) using the
+link on the landing page, or with the following:
+
+    curl -O 'https://data.nasa.gov/resource/gh4g-9sfh.json?$limit=50000'
+
+## Define cleansing functions
+
+The raw data for meteorite landings contains records that don't include a
+location or a year. These records are less useful because they fail to locate
+the meteorite in space-time, so it might make sense to remove them. Define a
+function that takes in the record, and yields it only if it has both a location
+and a year defined:
+
+[embedmd]:# (src/preprocessing/clean.py /def discard_incomplete/ /^$/)
+```py
+def discard_incomplete(record):
+    """Filters out records that don't have geolocation information."""
+    if 'geolocation' in record and 'year' in record:
+        yield record
+```
+
+The records in the source dataset have values that are always strings, but if
+you want to compare numerical fields as numbers, you'll need them to be
+interpreted as such. Define a function that casts the relevant fields to the
+appropriate types:
+
+[embedmd]:# (src/preprocessing/clean.py /def convert_types/ /return.*/)
+```py
+def convert_types(record):
+    """Converts string values to their appropriate type."""
+    # Only the year part of the datetime string is significant
+    record['year'] = int(record['year'][:4])
+
+    record['mass'] = float(record['mass']) if 'mass' in record else None
+
+    geolocation = record['geolocation']
+    geolocation['latitude'] = float(geolocation['latitude'])
+    geolocation['longitude'] = float(geolocation['longitude'])
+
+    return record
+```
+
+After having completed the preprocessing step and gone on to explore the data,
+you may notice that there are some suspicious values. For example, some entries
+have the value `0.0` for both `latitude` and `longitude`. Your explorations in
+the [exploratory section](bigquery.md) of the process will inform the filters
+you'll need to clean the data up - in this case, by discarding invalid values,
+and making educated guesses for the correct value when you can:
```

Comment:
It might be nice to be explicit that you're providing one example of potentially many value checks. And to make it clear that when you make an educated guess about what suspect values should be you may still end up with the wrong values (an educated guess still being a guess, after all).